standalone-metastore/metastore-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogActor.java

```diff
@@ -0,0 +1,274 @@
+import org.apache.hadoop.hive.metastore.api.LockResponse;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.ShowLocksRequest;
+import org.apache.hadoop.hive.metastore.api.ShowLocksResponse;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.api.UnlockRequest;
+import org.apache.hadoop.hive.metastore.utils.MetaStoreUtils;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.hive.HiveActor;
+import org.apache.iceberg.hive.HiveLock;
+import org.apache.iceberg.hive.MetastoreLock;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HMSCatalogActor implements HiveActor {
+  private static final Logger LOG = LoggerFactory.getLogger(HMSCatalogActor.class);
+  /** The actor name (catalog). */
+  private final String name;
+  /** The configuration (the Hadoop).  */
+  private final Configuration conf;
+  /** The client pool. */
+  private final ObjectPool<IHMSHandler> handlers;
+
+  private static IHMSHandler getHandler(Configuration configuration) throws MetaException {
+    IHMSHandler hmsHandler = new HMSHandler("HMSHandler", configuration);
+    try {
+      return HMSHandlerProxyFactory.getProxy(configuration, hmsHandler, true);
+    } catch (MetaException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public HMSCatalogActor(String name, Configuration configuration) {
+    this.name = name;
+    this.conf = configuration;
+    this.handlers = new GenericObjectPool<>(new BasePooledObjectFactory<IHMSHandler>() {
+      @Override
+      public IHMSHandler create() throws Exception {
+        return getHandler(new Configuration(conf));
+      }
+
+      @Override
+      public PooledObject<IHMSHandler> wrap(IHMSHandler ihmsHandler) {
+        return new DefaultPooledObject<>(ihmsHandler);
+      }
+    });
+  }
+
+  @FunctionalInterface
+  interface Action<R> {
+    R execute(IHMSHandler handler) throws TException;
+  }
+
+  private <R> R run(Action<R> action) throws TException {
+    IHMSHandler handler = null;
+    try {
+      try {
+        handler = handlers.borrowObject();
+      } catch (Exception e) {
+        throw new TException("run/borrowObject", e);
+      }
+      return action.execute(handler);
+    } finally {
+      if (handler != null) {
+        try {
+          handlers.returnObject(handler);
+        } catch (Exception e) {
+          LOG.error("run/returnObject", e);
+        }
+      }
+    }
+  }
+
+  @FunctionalInterface
+  interface VoidAction {
+    void execute(IHMSHandler handler) throws TException;
+  }
+
+  private void runVoid(VoidAction action) throws TException {
+    IHMSHandler handler = null;
+    try {
+      try {
```

Comment:
This double try seems strange. Could you please use just one level of try-catch-finally here? It seems unnecessary to complicate.